### PR TITLE
`cylc play` (restart): fix bug affecting run host reinvocation after interactive upgrade

### PR DESCRIPTION
**Pull request by MetRonnie**
Fixes a bug not recorded in an issue. I think this bug has been present since the introduction of the interactive upgrade mechanism (#5074)

TL;DR - remote reinvocation would not work after interactively upgrading a workflow to a new Cylc version on the CLI.

#### Steps to reproduce

With `run hosts` configured in `global.cylc`

```cylc
[scheduler]
    [[run hosts]]
        available = <server1>, ...
```

Start a workflow and stop it, then fiddle the workflow DB to make it look like it was run with an older version:

```
sqlite3 wflow/.service/db 'UPDATE workflow_params SET value="8.1.7" WHERE key="cylc_version"'
```

Now restart it with `cylc play`, and type 'y' at the interactive upgrade prompt. When it re-invokes on the run host, you should see an error

```
This workflow was previously run with 8.1.7.
This version of Cylc is 8.3.4.dev
Use "--upgrade" to upgrade the workflow.
ERROR: ERROR: command terminated by signal 1: ssh ...
```

#### Check List

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] No dependency changes 
- [x] Tests are included
- [x] Changelog entry included if this is a change that can affect users
- [x] No docs entry needed
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Comment by MetRonnie** [comment]
Test failures just macos functional issues and flaky integration test fixed by #6257

**Review by wxtim — APPROVED** [review]
👍🏼

File: cylc/flow/scheduler_cli.py [hunk 1]
```diff
@@ -612,8 +609,12 @@ def _distribute(host, workflow_id_raw, workflow_id, color):
         # Prevent recursive host selection
         cmd.append("--host=localhost")
 
+        # Ensure interactive upgrade carries over:
+        if options.upgrade and '--upgrade' not in cmd:
+            cmd.append('--upgrade')
```
Comment (oliver-sanders) [inline cylc/flow/scheduler_cli.py:614]: Needs duplicating for `--downgrade`?

File: cylc/flow/scheduler_cli.py [hunk 2]
```diff
@@ -390,11 +390,7 @@ async def scheduler_cli(
 
     # check the workflow can be safely restarted with this version of Cylc
     db_file = Path(get_workflow_srv_dir(workflow_id), 'db')
-    if not _version_check(
-        db_file,
-        options.upgrade,
-        options.downgrade,
-    ):
+    if not _version_check(db_file, options):
```
Comment (oliver-sanders) [inline cylc/flow/scheduler_cli.py:393]: FYI, I purposefully did *not* propagate the `options` object down the call chain. IMO it's a good idea to abstract out the CLI interfaces higher up in the call chain to keep the underlying code more abstract. The more you propagate CLI logic down the call stack, the harder it is to, for example, open a Python API later, or create a wrapper command, you also need to mock more in order to test and typing is less effective too (we know we're getting a `Values` object, but we don't know whether it has an `upgrade` option.

Not a blocker, but not positive "tidying" IMO.

File: cylc/flow/scheduler_cli.py (hunk 1 above)
Comment (MetRonnie) [reply cylc/flow/scheduler_cli.py:614]: There isn't an interactive downgrade though

Comment (oliver-sanders) [reply cylc/flow/scheduler_cli.py:614]: Gotcha.

**Review by oliver-sanders — APPROVED** [review]

File: cylc/flow/scheduler_cli.py (hunk 2 above)
Comment (MetRonnie) [reply cylc/flow/scheduler_cli.py:393]: I think ideally we want to only alter `options.upgrade` in the case of an interactive upgrade, so the simplest way of doing this is passing in the `options` object to `_version_check`. 

It is a shame about the type-looseness of it, but at least it is relatively easy to test these functions by importing and using `RunOptions` (no need for mocking).

https://github.com/cylc/cylc-flow/blob/6eb8f61c5b9777cb21ac68eff13aa4f0722f691a/cylc/flow/option_parsers.py#L569-L570

Comment (oliver-sanders) [reply cylc/flow/scheduler_cli.py:393]: `options.update = _version_check(db_file, options.upgrade, options.downgrade)` would have done it?

Comment (MetRonnie) [reply cylc/flow/scheduler_cli.py:393]: There are pathways it returns `True` without interactive upgrade
